File: sourcecode/gobalance/pkg/onionbalance/torNode.go
package onionbalance

import (
	"encoding/base64"
	"encoding/binary"
	"errors"
	"github.com/sirupsen/logrus"
	"golang.org/x/crypto/sha3"
	"strings"
	"sync"
)

type TorNode struct {
	microdescriptor    MicroDescriptor
	microdescriptorMtx sync.RWMutex
	routerstatus       *RouterStatus
	routerstatusMtx    sync.RWMutex
}

func NewNode(microdescriptor MicroDescriptor, routerstatus *RouterStatus) *TorNode {
	logrus.Debugf("Initializing node with fpr %s", routerstatus.Fingerprint)

	n := &TorNode{}
	n.setMicrodescriptor(microdescriptor)
	n.setRouterstatus(routerstatus)
	return n
}

func (n *TorNode) getRouterstatus() *RouterStatus {
	n.routerstatusMtx.RLock()
	defer n.routerstatusMtx.RUnlock()
	return n.routerstatus
}

func (n *TorNode) setRouterstatus(newVal *RouterStatus) {
	n.routerstatusMtx.Lock()
	defer n.routerstatusMtx.Unlock()
	n.routerstatus = newVal
}

func (n *TorNode) getMicrodescriptor() MicroDescriptor {
	n.microdescriptorMtx.RLock()
	defer n.microdescriptorMtx.RUnlock()
	return n.microdescriptor
}

func (n *TorNode) setMicrodescriptor(newVal MicroDescriptor) {
	n.microdescriptorMtx.Lock()
	defer n.microdescriptorMtx.Unlock()
	n.microdescriptor = newVal
}

func (n *TorNode) GetHexFingerprint() Fingerprint {
	return n.getRouterstatus().Fingerprint
}

var ErrNoHSDir = errors.New("NoHSDir")
var ErrNoEd25519Identity = errors.New("NoEd25519Identity")

// GetHsdirIndex get the HSDir index for this node:
//
//    hsdir_index(node) = H("node-idx" | node_identity |
//                          shared_random_value |
//                          INT_8(period_num) |
//                          INT_8(period_length) )
//
// Raises NoHSDir or NoEd25519Identity in case of errors.
func (n *TorNode) GetHsdirIndex(srv []byte, period_num int64, consensus *Consensus) ([]byte, error) {
	// See if this node can be an HSDir (it needs to be supported both in
	// protover and in flags)
	//arr, found := n.routerstatus.Protocols["HSDir"]
	//if !found {
	//	panic("NoHSDir")
	//}
	//found = false
	//for _, el := range arr {
	//	if 2 == el {
	//		found = true
	//		break
	//	}
	//}
	//if !found {
	//	panic("NoHSDir")
	//}
	if !n.getRouterstatus().Flags.HSDir {
		return nil, ErrNoHSDir
	}

	// See if ed25519 identity is supported for this node
	if _, found := n.getMicrodescriptor().Identifiers["ed25519"]; !found {
		return nil, ErrNoEd25519Identity
	}

	// In stem the ed25519 identity is a base64 string and we need to add
	// the missing padding so that the python base64 module can successfully
	// decode it.
	// TODO: Abstract this into its own function...
	ed25519NodeIdentityB64 := n.getMicrodescriptor().Identifiers["ed25519"]
	missingPadding := len(ed25519NodeIdentityB64) % 4
	ed25519NodeIdentityB64 += strings.Repeat("=", missingPadding)
	ed25519NodeIdentity, _ := base64.StdEncoding.DecodeString(ed25519NodeIdentityB64)

	periodNumInt8 := make([]byte, 8)
	binary.BigEndian.PutUint64(periodNumInt8[len(periodNumInt8)-8:], uint64(period_num))
	periodLength := consensus.Consensus().GetTimePeriodLength()
	periodLengthInt8 := make([]byte, 8)
	binary.BigEndian.PutUint64(periodLengthInt8[len(periodLengthInt8)-8:], uint64(periodLength))

	hashBody := "node-idx" + string(ed25519NodeIdentity) + string(srv) + string(periodNumInt8) + string(periodLengthInt8)
	hsdirIndex := sha3.Sum256([]byte(hashBody))

	return hsdirIndex[:], nil
}

File: sourcecode/rsbalance/src/onionbalance/tor_node.rs
use std::error::Error;
use std::fmt::{Display, Formatter};
use crate::onionbalance::consensus;
use crate::onionbalance::consensus::Consensus;
use crate::onionbalance::controller::MicroDescriptor;
use base64::{engine::general_purpose, Engine as _};
use sha3::Digest;
use sha3::Sha3_256;
use tor_netdoc::doc::netstatus::RouterStatus;
use tor_netdoc::doc::netstatus::{MdConsensusRouterStatus, RelayFlags};

#[derive(Clone)]
pub struct TorNode {
    microdescriptor: MicroDescriptor,
    pub routerstatus: MdConsensusRouterStatus,
}

impl TorNode {
    pub fn new(microdescriptor: MicroDescriptor, routerstatus: MdConsensusRouterStatus) -> Self {
        let tor_node = Self { microdescriptor, routerstatus };
        debug!("Initializing node with fpr {}", tor_node.get_hex_fingerprint());
        tor_node
    }
}

#[derive(Debug)]
pub enum TorNodeErr {
    ErrNoHSDir,
    ErrNoEd25519Identity,
}

impl Display for TorNodeErr {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}

impl Error for TorNodeErr {}

impl TorNode {
    pub fn get_hsdir_index(
        &self,
        srv: &Vec<u8>,
        period_num: u64,
        consensus: &Consensus,
    ) -> Result<[u8; 32], TorNodeErr> {
        // See if this node can be an HSDir (it needs to be supported both in
        // protover and in flags)
        if !self.routerstatus.flags().contains(RelayFlags::HSDIR) {
            return Err(TorNodeErr::ErrNoHSDir);
        }

        // See if ed25519 identity is supported for this node
        let Some(ed25519_value) = self.microdescriptor.identifiers.get("ed25519") else {
            return Err(TorNodeErr::ErrNoEd25519Identity);
        };

        // In stem the ed25519 identity is a base64 string and we need to add
        // the missing padding so that the python base64 module can successfully
        // decode it.
        // TODO: Abstract this into its own function...
        let mut ed25519_node_identity_b64 = ed25519_value.clone();
        let padding = (4 - ed25519_node_identity_b64.len() % 4) % 4;
        ed25519_node_identity_b64 += &"=".repeat(padding);
        let Ok(ed25519_node_identity) = general_purpose::STANDARD.decode(ed25519_node_identity_b64) else {
            return Err(TorNodeErr::ErrNoEd25519Identity);
        };
        let period_num_int8: [u8; 8] = period_num.to_be_bytes();
        let period_length = consensus.get_time_period_length();
        let period_length_int8: [u8; 8] = period_length.to_be_bytes();
        let mut hash_body = Vec::<u8>::new();
        hash_body.extend(b"node-idx");
        hash_body.extend(ed25519_node_identity);
        hash_body.extend(srv);
        hash_body.extend(period_num_int8);
        hash_body.extend(period_length_int8);
        let mut hasher = Sha3_256::new();
        hasher.update(hash_body);
        let hsdir_index: [u8; 32] = hasher.finalize().into();
        Ok(hsdir_index)
    }

    pub fn get_hex_fingerprint(&self) -> consensus::Fingerprint {
        let identity = &self.routerstatus.rsa_identity().to_string();
        crate::utils::strip_prefix(identity,"$").to_uppercase()
    }
}
